Redundancy documentation (#77)
ElrondNetwork/docs.elrond.com: docs/validators/redundancy.md

That configuration option called `RedundancyLevel` can be used as following:

- a 0 value will represent that the node is the main validator.
The value 0 will be the default, if the option is missing (as we do not re-write the `prefs.toml` 
  files during the node's upgrade) will still make that node the main validator: 
  backwards compatibility holds.
- a positive value will represent the "order of the back-up node" in the automatic fail-over sequence. 
  Example: let's suppose we have 3 nodes running with the same key. One with the redundancy level set to 0,
  one with 1 and one with 3. The node with level 0 will propose and sign upon blocks. The other 2 will 
  sync (and shuffle out) but will not sign anything. If the main validator fails, the back-up validator 
  with level 1 will start producing/signing blocks after `level*5` missed rounds. So, after 5 
  missed rounds by the main validator, the back-up node with level 1 will take the turn. 
  Now, if the back-up node 1 is down, the second back-up node will start producing/signing blocks 
  after `3*5 = 15 rounds` after the main validator failed and 10 rounds after the failed back-up node 1 

>   should have been produced a block.

Reviewer: Maybe one more proposition about the case when the main validator will come back after an offline period, all the redundancy nodes will become inactive again.

Author: done + added warning

---
id: redundancy
title: Node redundancy
---

Each validator node can be backed up by one or several backup nodes in order to 

> provide security of the network and to assure that a node will unlikely get jailed.

Reviewer 2: Elrond Validator Nodes can be configured to have one or more hot-standby nodes. This means additional nodes will run on different servers, in synch with the main Validator node. Their role is to stand in for the main Validator node in case it fails, to ensure high availability.

Author: Elrond Validator Nodes can be configured to have one or more hot-standby nodes. This means additional nodes will run on different servers, in **sync** with the main Validator node. Their role is to stand **by** for the main Validator node in case it fails, to ensure high availability. ?

The redundancy mechanism relies on the following principle: the owner of the BLS key will start `n` nodes
with almost the same configuration, and the same `validatorKey.pem` file. The difference between 

> configurations consists on an option inside the `prefs.toml` file.

Reviewer 2: Replace:
The redundancy mechanism relies on the following principle: the owner of the BLS key will start `n` nodes
with almost the same configuration, and the same `validatorKey.pem` file. The difference between 
configurations consists on an option inside the `prefs.toml` file.

with:
This is a redundancy mechanism which allows the Main Validator operator to start additional 'n' hot-standby nodes, each of them running the same 'validatorKey.pem' file. 

Hot standby nodes are configured using the 'RedundancyLevel' option in the 'prefs.toml' configuration file.

Author: done

> That configuration option called `RedundancyLevel` can be used as following:

Reviewer 2: remove this line as well

Author: done

> - a 0 value will represent that the node is the main validator.

Reviewer 2: main validator -> Main Validator

Author: done

The value 0 will be the default, if the option is missing (as we do not re-write the `prefs.toml` 
  files during the node's upgrade) will still make that node the main validator: 
  backwards compatibility holds.

> - a positive value will represent the "order of the back-up node" in the automatic fail-over sequence. 

Reviewer 2: back-up node -> hot-standby node

Author: done

  Example: let's suppose we have 3 nodes running with the same key. One with the redundancy level set to 0,

>   one with 1 and one with 3. The node with level 0 will propose and sign upon blocks. The other 2 will 

Reviewer 2: remove "upon"

Author: done

>   sync (and shuffle out) but will not sign anything. If the main validator fails, the back-up validator 

Reviewer 2: replace:
sync (and shuffle out)

with:
sync data with the same shard as the Main Validator (and shuffle in and out of the same shards)

Reviewer 2: Replace:

"back-up validator"

with "hot-standby node"

Author: done and done

  with level 1 will start producing/signing blocks after `level*5` missed rounds. So, after 5 

>   missed rounds by the main validator, the back-up node with level 1 will take the turn. 

Reviewer 2: main validator -> Main Validator

back-up node -> hot-standby node"

Author: done and done

>   Now, if the back-up node 1 is down, the second back-up node will start producing/signing blocks 

Reviewer 2: Replace:
Now, if the back-up node 1 is down, the second back-up node will start producing/signing blocks 

with
If hot-standby node 1 is down as well, hot-standby node 2 will step in

Author: done

>   after `3*5 = 15 rounds` after the main validator failed and 10 rounds after the failed back-up node 1 

Reviewer 2: main validator -> Main Validator

back-up node -> hot-standby node

Author: done and done

  should have been produced a block.
- a large value for this level option (say 1 million), or a negative value (say -1) will mean that the 

>   backup observers won't get the chance to produce/sign blocks but will sync with the network and 

Reviewer 2: backup observers -> hot-standby nodes

Author: done

>   shuffle out just as the main validator will behave.

Reviewer 2: shuffle out just as the main validator will behave. -> shuffle between shards just as the Main Validator will.

Author: done

:::tip

> The backup nodes will advertise on the network a different public key (autogenerated at start-up) and thus, concealing the real public key that will be used when signing the header blocks. 

Reviewer 2: backup nodes -> hot-standby nodes

Author: done

:::

:::tip

> If a validator recovers and start signing blocks again, all other validators with higher `RedundancyLevel` will become inactive again.

Reviewer 2: Replace:

If a validator recovers and start signing blocks again, all other validators with higher `RedundancyLevel` will become inactive again.

with:

If the Main Validator (RedundancyLevel 0) gets back online, the hot-standby node(s) revert to standby mode.

Author: Partially true, if main validator is still offline but hot-standby 1 is online, only hot-standby 2, 3....n nodes will become inactive

:::

:::warning
Do not use the same redundancy level on more than one node as the nodes with the same `RedundancyLevel` value will start signing blocks in parallel in the same time. The protocol can withstand this kind of things safely but in the near future the BLS key that will perform double signing can get slashed.
:::

The main reasons for using another BLS key are the following: 

> - the back-up node(s) will not cause BLS signature re-verification when idle.

Reviewer 2: replace: The main reasons for using another BLS key are the following: 

with: The random BLS key on hot-standby nodes has the following purposes:

Reviewer 2: back-up node -> hot-standby node

Author: done and done

- it slightly prevents DDoS attacks as an attacker can not find all IPs behind a targeted BLS public key: 

>   when an attacker takes down the main host, the back-up nodes will advertise the public key when they 

Reviewer 2: main host -> Main Validator
back-up nodes -> hot-standby nodes

Author: done and done

---
id: redundancy
title: Node redundancy
---

Elrond Validator Nodes can be configured to have one or more hot-standby nodes. 
This means additional nodes will run on different servers, in sync with the Main Validator node. 
Their role is to stand in for the Main Validator node in case it fails, to ensure high availability.

This is a redundancy mechanism which allows the Main Validator operator to start additional 'n' hot-standby nodes, 
each of them running the same 'validatorKey.pem' file. The difference between 
configurations consists on an option inside the `prefs.toml` file.

Hot standby nodes are configured using the 'RedundancyLevel' option in the 'prefs.toml' configuration file:

- a 0 value will represent that the node is the Main Validator.

> The value 0 will be the default, if the option is missing (as we do not re-write the `prefs.toml` 
>   files during the node's upgrade) will still make that node the Main Validator: 
>   backwards compatibility holds.

Reviewer 3: ```suggestion
The value 0 will be the default, therefore if the option is missing it will still make that node the Main Validator by default. With consideration to backwards compatibility, the already-running Validators are not affected by the addition of this option. Moreover, we never overwrite the `prefs.toml` files during the node's upgrade.

The values of `RedundancyLevel` are interpreted as follows:
```

- a positive value will represent the "order of the hot-standby node" in the automatic fail-over sequence. 

>   Example: let's suppose we have 3 nodes running with the same key. One with the redundancy level set to 0,

Reviewer 3: ```suggestion
  Example: suppose we have 3 nodes running with the same BLS key. One has the redundancy level set to 0,
```

>   one with 1 and one with 3. The node with level 0 will propose and sign blocks. The other 2 will

Reviewer 3: ```suggestion
  another has 1 and another with 3. The node with level 0 will propose and sign blocks. The other 2 will
```

  sync data with the same shard as the Main Validator (and shuffle in and out of the same shards) but will
  not sign anything. If the Main Validator fails, the hot-standby node 
  with level 1 will start producing/signing blocks after `level*5` missed rounds. So, after 5 
  missed rounds by the Main Validator, the hot-standby node with level 1 will take the turn.
  If hot-standby node 1 is down as well, hot-standby node 2 will step in 
  after `3*5 = 15 rounds` after the Main Validator failed and 10 rounds after the failed hot-standby node 1 
  should have been produced a block.
- a large value for this level option (say 1 million), or a negative value (say -1) will mean that the
  hot-standby nodes won't get the chance to produce/sign blocks but will sync with the network and
  shuffle between shards just as the Main Validator will.

:::tip
The hot-standby nodes will advertise on the network a different public key (autogenerated at start-up) and thus, concealing the real public key that will be used when signing the header blocks. 
:::

:::tip
If the Main Validator (RedundancyLevel 0) gets back online, the hot-standby node(s) revert to standby mode.
:::

:::warning

> Do not use the same redundancy level on more than one node as the nodes with the same `RedundancyLevel` value will start signing blocks in parallel in the same time. The protocol can withstand this kind of things safely but in the near future the BLS key that will perform double signing can get slashed.

Reviewer 3: ```suggestion
Do not use the same redundancy level on more than one node. Otherwise, the nodes with the same `RedundancyLevel` value will start signing blocks in parallel in the same time. Although the protocol is not negatively affected by double signing, in the near future the BLS key that will perform double signing will have its stake slashed.
```

:::

The random BLS key on hot-standby nodes has the following purposes:

> - the hot-standby node(s) will not cause BLS signature re-verification when idle.

Reviewer 3: Might need some clarification as to what "BLS signature re-verification" means.

Author: I think we can let it as it is: the mechanism is way to complicated to explain in this section.

> - it slightly prevents DDoS attacks as an attacker can not find all IPs behind a targeted BLS public key: 

Reviewer 3: ```suggestion
- it helps preventing DDoS attacks because an attacker cannot find all IPs behind a targeted BLS public key: 
```

>   when an attacker takes down the Main Validator, the hot-standby nodes will advertise the public key when they 

Reviewer 3: ```suggestion
  when an attacker takes down the Main Validator, only then will the hot-standby nodes advertise the public key 
```

>   will start signing blocks (not sooner).

Reviewer 3: ```suggestion
  by using it to sign blocks, but not sooner.
```
